File: programmers_problmes/python3/단어변환_lv3.py
answer = 1000000

def check(word1, word2):
    diff = 0
    for i in range(len(word1)):
        if word1[i] != word2[i]:
            diff += 1
            
    if diff == 1:
        return 1
    else:
        return 0
            
            
def dfs(word, words, target, count, visit):
    global answer
    if word == target:
        answer = min(answer, count)
        return
    
    if len(visit) == len(words):
        return
    
    for i in range(len(words)):
        if check(word, words[i]) and words[i] not in visit:
            visit.append(words[i])
            dfs(words[i], words, target, count+1, visit)
            visit.pop()
    
    

def solution(begin, target, words):
    global answer
    
    if target not in words:
        return 0
    
    dfs(begin, words, target, 0, [])

    return answer

File: programmers_problmes/kotlin/단어변환_lv3.kt
import kotlin.math.*

class Solution {
    var answer = 0

    fun solution(begin: String, target: String, words: Array<String>): Int {


        dfs(words, MutableList(words.size){false}, begin, target, 0, mutableListOf<String>())

        return answer
    }


    fun dfs(words: Array<String>, visit: MutableList<Boolean>, begin: String, target: String, count: Int, wordList: MutableList<String>) {


        if(begin == target) {
            if(answer == 0) answer = wordList.size
            else answer = min(answer, wordList.size)
        }

        for(i in 0 until words.size) {
            if(visit[i] == false && checkIfStringHasOneCharaterDiff(begin, words[i])) {

                visit[i] = true
                wordList.add(words[i])

                dfs(words, visit, words[i], target, count+1, wordList)

                visit[i] = false
                wordList.removeAt(wordList.lastIndex)
            }
        }

    }

    fun checkIfStringHasOneCharaterDiff(word1: String, word2: String): Boolean {
        var diffCount = 0

        for(i in word1.indices) {
            if(word1[i] != word2[i]) diffCount++
        }

        if(diffCount == 1) return true
        else return false

    }



}